### PR TITLE
[DEV-6458] Paginated and API Contract Integrated Dates & Details Table

### PR DESCRIPTION
**Pull request by MaxwellKendall**
**High level description:**
(Contains my other pr #2264)

Dates and Details Page are:

(a) paginated & sortable
(b) hooked up to api locally & when hosted using mock data from contract
(c) using models

**Mock Up**
Dates Table: https://bahdigital.invisionapp.com/share/4PIAFJCDK5Q#/screens/296052481
Details Table: https://bahdigital.invisionapp.com/share/4PIAFJCDK5Q#/screens/296052480

**Technical details:**
- Implementing the `redux.aboutTheData` namespace (1dab82c tests: bd4ae03)
- Creating mock data for `totalsRequest`, `datesRequest` and `detailsRequest` f34b081
- Invoking the API w/o pagination 8da0841 & adding styles for not certified badge 👇 
image
- pagination and styling d727bcf

TESTS:
- 4b3fa58 ensuring container does not make duplicate requests; discovered multiple bugs related to this and also one related to fetching an invalid period (for example, period 7 is only selectable in 2020, so we added some validation to our request such that if user is on period 7, FY 2020, then changes to FY 2018, we don't fetch period 7 2018)
- b4c9006 redux
- 43bf78d model (NOTE: significant overlap w/ @ebdabbs implementation on her latest PR for the agency detail page. Likely will amend this commit after she's merged to ensure total coverage on the model used for the details table)

**JIRA Ticket:**
[DEV-6458](https://federal-spending-transparency.atlassian.net/browse/DEV-6458)

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [x] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
`N/A` Verified cross-browser compatibility: Chrome, Safari, Firefox, Internet Explorer, Edge
- [x] Verified mobile/tablet/desktop/monitor responsiveness
`N/A` Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for helper functions, reducers, models and Container/Component Interactivity Expectations `if applicable` [React Testing Library](react-testing-library.md)
`N/A` [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`
`N/A` [Component Library Integration Status Report](https://github.com/fedspendingtransparency/data-act-documentation/blob/data-transparency-ui/frontend_apps/component-library-integration-status.md) updated `if applicable`

Reviewer(s):
- [x] Design review complete `if applicable`
`N/A` [API #1234](https://github.com/fedspendingtransparency/usaspending-api/pull/1234) merged concurrently `if applicable`
- [ ] Code review complete

### CODE REVIEW

File: src/js/containers/aboutTheData/AgenciesContainer.jsx [hunk 1]
```diff
@@ -1,93 +1,276 @@
...
+    const dispatch = useDispatch();
+    const datesReq = useRef(null);
+    const detailsReq = useRef(null);
+    const totalsReq = useRef(null);
+    const [
+        [{ page: detailsPage, limit: detailsLimit }, updateDetailsPagination],
+        [{ page: datesPage, limit: datesLimit }, updateDatesPagination]
+    ] = [usePagination(), usePagination()];
+    const [{ vertical: isVerticalSticky, horizontal: isHorizontalSticky }, setIsSticky] = useState({ vertical: false, horizontal: false });
+    const [[, areDetailsLoading, areDatesLoading], setLoading] = useState([true, true, true]);
```
Comment (jonhill13) [inline src/js/containers/aboutTheData/AgenciesContainer.jsx:51]: Do we want the first item of this array to be empty?

File: src/js/helpers/aboutTheDataHelper.js [hunk 2]
```diff
@@ -2,10 +2,65 @@
...
+// periods are only selectable post 2020
+export const isPeriodSelectable = (availablePeriodsInFy, periodId) => (
+    availablePeriodsInFy
+        .filter((p) => (
+            parseInt(periodId, 10) === p.submission_fiscal_month
+        ))
+        .length > 0
+);
+
+// getting last period of quarter for period via index of this array. ✨✨ ✨  S/O to (3rd grade) Maths ✨ ✨ ✨
```
Comment (jonhill13) [inline src/js/helpers/aboutTheDataHelper.js:59]: Maths should not be here, maybe Math? or Mathematics

File: src/js/helpers/aboutTheDataHelper.js [hunk 3]
```diff
@@ -18,6 +73,101 @@ export const aboutTheDataQueryString = (params) => {
...
+    page: 1,
+    limit: 10
+};
+
+export const usePagination = (initialState = defaultState) => {
+    const [{ page, limit }, updatePagination] = useState(initialState);
+    return [{ page, limit }, updatePagination];
+};
+
+export const getTotals = (fy = '2020', period = '12') => {
```
Comment (jonhill13) [inline src/js/helpers/aboutTheDataHelper.js:86]: we should follow our convention for api calls and be more descriptive `fetchTotalBudgetaryResources`

File: src/js/helpers/aboutTheDataHelper.js [hunk 4]
```diff
@@ -2,10 +2,65 @@
  * aboutTheDataHelper.js
  * Created by Jonathan Hill 11/20/20
  */
+import { useState } from 'react';
+import { stringify } from 'query-string';
```
Comment (jonhill13) [inline src/js/helpers/aboutTheDataHelper.js:6]: This library should be removed from our package.json. Node has it's own query string module.

https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options

**Review by jonhill13 — CHANGES_REQUESTED** [review]

File: src/js/redux/reducers/aboutTheData/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/**
+ * aboutTheDataReducer.js
+ * Created by Max Kendall 12/8/2020
+*/
+
+export const initialState = {
+    details: [],
+    dates: [],
```
Comment (jonhill13) [inline src/js/redux/reducers/aboutTheData/index.js:8]: I am assuming dates is the periods response? This also should be more descriptive.

**Review by jonhill13 — CHANGES_REQUESTED** [review]

File: src/js/containers/aboutTheData/AgenciesContainer.jsx (hunk 1 above)
Comment (MaxwellKendall) [reply src/js/containers/aboutTheData/AgenciesContainer.jsx:51]: the first item, yes

File: src/_scss/pages/aboutTheData/agenciesPage.scss [hunk 6]
```diff
@@ -40,9 +39,170 @@
...
+        padding-left: 0;
+        @media (min-width: $medium-screen) {
+          padding-left: rem(25);
+        }
+        margin-top: rem(25);
+        @include display(flex);
+        @include justify-content(flex-start);
+        @include align-items(flex-start);
+        .usa-dt-picker__button-icon svg {
+          color:  #0074BE;
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply src/_scss/pages/aboutTheData/agenciesPage.scss:55]: nope 
image

File: src/_scss/pages/aboutTheData/agenciesPage.scss [hunk 7]
```diff
@@ -81,8 +242,19 @@
               font-size: 1.4rem;
               padding: 0; // to allow entire cell to be hoverable
               @import './actionCell';
+              flex-wrap:nowrap;
               .generic-cell-content {
                 padding: 1.2rem;
+                .not-certified {
+                  margin-left: rem(5);
+                  background: #DCE4EE;
```
[the remark this answers is not in the document]
Comment (MaxwellKendall) [reply src/_scss/pages/aboutTheData/agenciesPage.scss:250]: nope

File: src/js/redux/reducers/aboutTheData/index.js (hunk 5 above)
Comment (MaxwellKendall) [reply src/js/redux/reducers/aboutTheData/index.js:8]: This is one i disagree w/ b/c throughout the code I refer to the table as the `dates` table. This is already nested inside `aboutTheData`. I can change in a follow up clean up pr at some point if necessary.

File: src/_scss/pages/aboutTheData/agenciesPage.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonhill13) [reply src/_scss/pages/aboutTheData/agenciesPage.scss:55]: So we should change these to a variable from our color scheme

File: src/_scss/pages/aboutTheData/agenciesPage.scss [hunk 8]
```diff
@@ -40,9 +39,170 @@
...
+          }
+        }
+        .picker-container {
+          @include display(flex);
+          @include flex-direction(column);
+          .fy-picker__title,
+          .period-picker__title {
+            font-size: rem(14);
+            margin-left: rem(15);
+            color: #8C8C8C;
```
[the remark this answers is not in the document]
Comment (jonhill13) [reply src/_scss/pages/aboutTheData/agenciesPage.scss:84]: So we should change these to a variable from our color scheme

File: src/_scss/pages/aboutTheData/agenciesPage.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonhill13) [reply src/_scss/pages/aboutTheData/agenciesPage.scss:250]: So we should change these to a variable from our color scheme

**Review by jonhill13 — APPROVED** [review]

**Comment by MaxwellKendall** [comment]
Outstanding change requests noted.

**Pull request merged**